File: android/app/src/main/java/com/adblock/analytics/Analytics.kt
package com.adblock.analytics

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import kotlinx.coroutines.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * Privacy-focused analytics for AdBlock
 * Only collects anonymous usage data to improve the app
 */
class Analytics(private val context: Context) {
    companion object {
        private const val TAG = "Analytics"
        private const val PREFS_NAME = "adblock_analytics"
        private const val KEY_ANONYMOUS_ID = "anonymous_id"
        private const val KEY_ENABLED = "analytics_enabled"
        private const val MAX_EVENTS = 1000
        
        @Volatile
        private var instance: Analytics? = null
        
        fun getInstance(context: Context): Analytics {
            return instance ?: synchronized(this) {
                instance ?: Analytics(context.applicationContext).also {
                    instance = it
                }
            }
        }
    }
    
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val events = mutableListOf<AnalyticsEvent>()
    private val metrics = ConcurrentHashMap<String, MetricValue>()
    private val coroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val json = Json { ignoreUnknownKeys = true }
    
    private var sessionId = UUID.randomUUID().toString()
    private var sessionStartTime = System.currentTimeMillis()
    private var lastActivityTime = System.currentTimeMillis()
    
    val anonymousId: String = prefs.getString(KEY_ANONYMOUS_ID, null) ?: run {
        val id = UUID.randomUUID().toString()
        prefs.edit().putString(KEY_ANONYMOUS_ID, id).apply()
        id
    }
    
    var isEnabled: Boolean
        get() = prefs.getBoolean(KEY_ENABLED, true)
        set(value) {
            prefs.edit().putBoolean(KEY_ENABLED, value).apply()
            if (!value) {
                clearAllData()
            }
        }
    
    /**
     * Track an event
     */
    @JvmOverloads
    fun trackEvent(
        name: String,
        category: EventCategory,
        properties: Map<String, Any> = emptyMap()
    ) {
        if (!isEnabled) return
        
        synchronized(events) {
            if (events.size >= MAX_EVENTS) {
                events.removeAt(0)
            }
            
            events.add(AnalyticsEvent(
                name = name,
                category = category,
                properties = properties,
                timestamp = System.currentTimeMillis(),
                sessionId = sessionId
            ))
        }
        
        lastActivityTime = System.currentTimeMillis()
    }
    
    /**
     * Track a simple action
     */
    fun trackAction(action: String) {
        trackEvent(action, EventCategory.ACTION)
    }
    
    /**
     * Track feature usage
     */
    fun trackFeature(feature: String, properties: Map<String, Any> = emptyMap()) {
        trackEvent(feature, EventCategory.FEATURE, properties)
    }
    
    /**
     * Track performance metric
     */
    fun trackPerformance(metric: String, value: Double) {
        trackEvent(metric, EventCategory.PERFORMANCE, mapOf("value" to value))
        recordMetric(metric, value)
    }
    
    /**
     * Track error (non-crash)
     */
    fun trackError(error: String, errorType: String) {
        trackEvent(error, EventCategory.ERROR, mapOf("error_type" to errorType))
    }
    
    /**
     * Record a metric value
     */
    fun recordMetric(name: String, value: Double) {
        if (!isEnabled) return
        
        metrics.compute(name) { _, existing ->
            when (existing) {
                is MetricValue.Count -> MetricValue.Count(existing.value + 1)
                is MetricValue.Sum -> MetricValue.Sum(existing.value + value)
                is MetricValue.Average -> MetricValue.Average(
                    sum = existing.sum + value,
                    count = existing.count + 1
                )
                is MetricValue.Distribution -> {
                    val values = existing.values.toMutableList()
                    values.add(value)
                    if (values.size > 1000) {
                        values.subList(0, 500).clear()
                    }
                    MetricValue.Distribution(values)
                }
                null -> MetricValue.Sum(value)
            }
        }
    }
    
    /**
     * Increment a counter
     */
    fun incrementCounter(name: String) {
        metrics.compute(name) { _, existing ->
            when (existing) {
                is MetricValue.Count -> MetricValue.Count(existing.value + 1)
                else -> MetricValue.Count(1)
            }
        }
    }
    
    /**
     * Start a new session
     */
    fun startSession() {
        sessionId = UUID.randomUUID().toString()
        sessionStartTime = System.currentTimeMillis()
        lastActivityTime = System.currentTimeMillis()
        
        trackEvent("session_start", EventCategory.LIFECYCLE)
    }
    
    /**
     * End the current session
     */
    fun endSession() {
        val duration = (lastActivityTime - sessionStartTime) / 1000
        trackEvent("session_end", EventCategory.LIFECYCLE, mapOf("duration_seconds" to duration))
    }
    
    /**
     * Get analytics summary
     */
    fun getSummary(): AnalyticsSummary {
        val eventsByCategory = events.groupingBy { it.category }.eachCount()
        val metricSummaries = metrics.mapValues { (_, value) ->
            when (value) {
                is MetricValue.Count -> "Count: ${value.value}"
                is MetricValue.Sum -> "Sum: %.2f".format(value.value)
                is MetricValue.Average -> {
                    val avg = if (value.count > 0) value.sum / value.count else 0.0
                    "Avg: %.2f (n=${value.count})".format(avg)
                }
                is MetricValue.Distribution -> {
                    if (value.values.isEmpty()) {
                        "No data"
                    } else {
                        val min = value.values.minOrNull() ?: 0.0
                        val max = value.values.maxOrNull() ?: 0.0
                        val avg = value.values.average()
                        "Min: %.2f, Max: %.2f, Avg: %.2f".format(min, max, avg)
                    }
                }
            }
        }
        
        return AnalyticsSummary(
            totalEvents = events.size,
            eventsByCategory = eventsByCategory,
            metrics = metricSummaries,
            sessionDuration = (System.currentTimeMillis() - sessionStartTime) / 1000
        )
    }
    
    /**
     * Export events for analysis
     */
    fun exportEvents(limit: Int = 100): List<AnalyticsEvent> {
        return synchronized(events) {
            events.takeLast(limit).toList()
        }
    }
    
    /**
     * Clear all analytics data
     */
    fun clearAllData() {
        synchronized(events) {
            events.clear()
        }
        metrics.clear()
    }
    
    // Pre-defined event tracking methods
    
    fun trackAppLaunch(launchTimeMs: Long) {
        trackEvent("app_launch", EventCategory.LIFECYCLE, mapOf("launch_time_ms" to launchTimeMs))
    }
    
    fun trackVpnConnected(connectionTimeMs: Long) {
        trackEvent("vpn_connected", EventCategory.ACTION, mapOf("connection_time_ms" to connectionTimeMs))
        incrementCounter("vpn_connections")
    }
    
    fun trackVpnDisconnected(reason: String) {
        trackEvent("vpn_disconnected", EventCategory.ACTION, mapOf("reason" to reason))
    }
    
    fun trackFilterUpdated(rulesCount: Int, durationMs: Long) {
        trackEvent("filter_updated", EventCategory.ACTION, mapOf(
            "rules_count" to rulesCount,
            "duration_ms" to durationMs
        ))
    }
    
    fun trackCustomRuleAdded(ruleType: String) {
        trackEvent("custom_rule_added", EventCategory.FEATURE, mapOf("rule_type" to ruleType))
        incrementCounter("custom_rules")
    }
    
    fun trackAdBlocked(sizeBytes: Long) {
        // Don't track domains for privacy
        incrementCounter("ads_blocked")
        recordMetric("bytes_saved", sizeBytes.toDouble())
    }
    
    fun trackPerformanceWarning(metric: String, value: Double) {
        trackEvent("performance_warning", EventCategory.PERFORMANCE, mapOf(
            "metric" to metric,
            "value" to value
        ))
    }
}

/**
 * Analytics event
 */
@Serializable
data class AnalyticsEvent(
    val name: String,
    val category: EventCategory,
    val properties: Map<String, Any>,
    val timestamp: Long,
    val sessionId: String
)

/**
 * Event categories
 */
@Serializable
enum class EventCategory {
    LIFECYCLE,
    ACTION,
    PERFORMANCE,
    ERROR,
    FEATURE
}

/**
 * Metric value types
 */
sealed class MetricValue {
    data class Count(val value: Long) : MetricValue()
    data class Sum(val value: Double) : MetricValue()
    data class Average(val sum: Double, val count: Long) : MetricValue()
    data class Distribution(val values: List<Double>) : MetricValue()
}

/**
 * Analytics summary
 */
data class AnalyticsSummary(
    val totalEvents: Int,
    val eventsByCategory: Map<EventCategory, Int>,
    val metrics: Map<String, String>,
    val sessionDuration: Long
)

File: ios/AdBlock/Analytics/Analytics.swift
import Foundation
import UIKit

/// Privacy-focused analytics for AdBlock
/// Only collects anonymous usage data to improve the app
public class Analytics {
    
    // MARK: - Properties
    
    static let shared = Analytics()
    
    private let analyticsQueue = DispatchQueue(label: "com.adblock.analytics", qos: .background)
    private let userDefaults = UserDefaults.standard
    private var events: [AnalyticsEvent] = []
    private var metrics: [String: MetricValue] = [:]
    private let maxEvents = 1000
    
    private var sessionId = UUID().uuidString
    private var sessionStartTime = Date()
    private var lastActivityTime = Date()
    
    public let anonymousId: String
    
    public var isEnabled: Bool {
        get { userDefaults.bool(forKey: "analytics_enabled") }
        set {
            userDefaults.set(newValue, forKey: "analytics_enabled")
            if !newValue {
                clearAllData()
            }
        }
    }
    
    // MARK: - Initialization
    
    private init() {
        // Get or create anonymous ID
        if let savedId = userDefaults.string(forKey: "anonymous_id") {
            self.anonymousId = savedId
        } else {
            self.anonymousId = UUID().uuidString
            userDefaults.set(anonymousId, forKey: "anonymous_id")
        }
        
        // Set default enabled state
        if userDefaults.object(forKey: "analytics_enabled") == nil {
            userDefaults.set(true, forKey: "analytics_enabled")
        }
        
        // Listen for app lifecycle events
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appWillResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )
    }
    
    // MARK: - Public Methods
    
    /// Track an event
    public func trackEvent(
        _ name: String,
        category: EventCategory,
        properties: [String: Any] = [:]
    ) {
        guard isEnabled else { return }
        
        analyticsQueue.async { [weak self] in
            guard let self = self else { return }
            
            let event = AnalyticsEvent(
                name: name,
                category: category,
                properties: properties,
                timestamp: Date(),
                sessionId: self.sessionId
            )
            
            self.events.append(event)
            
            // Limit number of events
            if self.events.count > self.maxEvents {
                self.events.removeFirst(self.events.count - self.maxEvents)
            }
            
            self.lastActivityTime = Date()
        }
    }
    
    /// Track a simple action
    public func trackAction(_ action: String) {
        trackEvent(action, category: .action)
    }
    
    /// Track feature usage
    public func trackFeature(_ feature: String, properties: [String: Any] = [:]) {
        trackEvent(feature, category: .feature, properties: properties)
    }
    
    /// Track performance metric
    public func trackPerformance(_ metric: String, value: Double) {
        trackEvent(metric, category: .performance, properties: ["value": value])
        recordMetric(metric, value: value)
    }
    
    /// Track error (non-crash)
    public func trackError(_ error: String, errorType: String) {
        trackEvent(error, category: .error, properties: ["error_type": errorType])
    }
    
    /// Record a metric value
    public func recordMetric(_ name: String, value: Double) {
        guard isEnabled else { return }
        
        analyticsQueue.async { [weak self] in
            guard let self = self else { return }
            
            if let existing = self.metrics[name] {
                switch existing {
                case .count(let count):
                    self.metrics[name] = .count(count + 1)
                case .sum(let sum):
                    self.metrics[name] = .sum(sum + value)
                case .average(let sum, let count):
                    self.metrics[name] = .average(sum: sum + value, count: count + 1)
                case .distribution(var values):
                    values.append(value)
                    if values.count > 1000 {
                        values.removeFirst(500)
                    }
                    self.metrics[name] = .distribution(values)
                }
            } else {
                self.metrics[name] = .sum(value)
            }
        }
    }
    
    /// Increment a counter
    public func incrementCounter(_ name: String) {
        guard isEnabled else { return }
        
        analyticsQueue.async { [weak self] in
            guard let self = self else { return }
            
            if case .count(let count) = self.metrics[name] {
                self.metrics[name] = .count(count + 1)
            } else {
                self.metrics[name] = .count(1)
            }
        }
    }
    
    /// Start a new session
    public func startSession() {
        sessionId = UUID().uuidString
        sessionStartTime = Date()
        lastActivityTime = Date()
        
        trackEvent("session_start", category: .lifecycle)
    }
    
    /// End the current session
    public func endSession() {
        let duration = lastActivityTime.timeIntervalSince(sessionStartTime)
        trackEvent("session_end", category: .lifecycle, properties: ["duration_seconds": Int(duration)])
    }
    
    /// Get analytics summary
    public func getSummary() -> AnalyticsSummary {
        var summary = AnalyticsSummary()
        
        analyticsQueue.sync {
            // Count events by category
            summary.eventsByCategory = Dictionary(grouping: events, by: { $0.category })
                .mapValues { $0.count }
            
            // Summarize metrics
            summary.metrics = metrics.compactMapValues { value in
                switch value {
                case .count(let count):
                    return "Count: \(count)"
                case .sum(let sum):
                    return "Sum: \(String(format: "%.2f", sum))"
                case .average(let sum, let count):
                    let avg = count > 0 ? sum / Double(count) : 0.0
                    return "Avg: \(String(format: "%.2f", avg)) (n=\(count))"
                case .distribution(let values):
                    if values.isEmpty {
                        return "No data"
                    } else {
                        let min = values.min() ?? 0
                        let max = values.max() ?? 0
                        let avg = values.reduce(0, +) / Double(values.count)
                        return "Min: \(String(format: "%.2f", min)), Max: \(String(format: "%.2f", max)), Avg: \(String(format: "%.2f", avg))"
                    }
                }
            }
            
            summary.totalEvents = events.count
            summary.sessionDuration = Int(Date().timeIntervalSince(sessionStartTime))
        }
        
        return summary
    }
    
    /// Export events for analysis
    public func exportEvents(limit: Int = 100) -> [AnalyticsEvent] {
        return analyticsQueue.sync {
            Array(events.suffix(limit))
        }
    }
    
    /// Clear all analytics data
    public func clearAllData() {
        analyticsQueue.async { [weak self] in
            self?.events.removeAll()
            self?.metrics.removeAll()
        }
    }
    
    // MARK: - Pre-defined Events
    
    public func trackAppLaunch(launchTimeMs: Int64) {
        trackEvent("app_launch", category: .lifecycle, properties: ["launch_time_ms": launchTimeMs])
    }
    
    public func trackVpnConnected(connectionTimeMs: Int64) {
        trackEvent("vpn_connected", category: .action, properties: ["connection_time_ms": connectionTimeMs])
        incrementCounter("vpn_connections")
    }
    
    public func trackVpnDisconnected(reason: String) {
        trackEvent("vpn_disconnected", category: .action, properties: ["reason": reason])
    }
    
    public func trackFilterUpdated(rulesCount: Int, durationMs: Int64) {
        trackEvent("filter_updated", category: .action, properties: [
            "rules_count": rulesCount,
            "duration_ms": durationMs
        ])
    }
    
    public func trackCustomRuleAdded(ruleType: String) {
        trackEvent("custom_rule_added", category: .feature, properties: ["rule_type": ruleType])
        incrementCounter("custom_rules")
    }
    
    public func trackAdBlocked(sizeBytes: Int64) {
        // Don't track domains for privacy
        incrementCounter("ads_blocked")
        recordMetric("bytes_saved", value: Double(sizeBytes))
    }
    
    public func trackPerformanceWarning(metric: String, value: Double) {
        trackEvent("performance_warning", category: .performance, properties: [
            "metric": metric,
            "value": value
        ])
    }
    
    // MARK: - Private Methods
    
    @objc private func appDidBecomeActive() {
        startSession()
    }
    
    @objc private func appWillResignActive() {
        endSession()
    }
}

// MARK: - Data Models

/// Analytics event
public struct AnalyticsEvent: Codable {
    let name: String
    let category: EventCategory
    let properties: [String: Any]
    let timestamp: Date
    let sessionId: String
    
    enum CodingKeys: String, CodingKey {
        case name, category, timestamp, sessionId
    }
    
    public init(name: String, category: EventCategory, properties: [String: Any], timestamp: Date, sessionId: String) {
        self.name = name
        self.category = category
        self.properties = properties
        self.timestamp = timestamp
        self.sessionId = sessionId
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        name = try container.decode(String.self, forKey: .name)
        category = try container.decode(EventCategory.self, forKey: .category)
        timestamp = try container.decode(Date.self, forKey: .timestamp)
        sessionId = try container.decode(String.self, forKey: .sessionId)
        properties = [:] // Properties not decoded for simplicity
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(name, forKey: .name)
        try container.encode(category, forKey: .category)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(sessionId, forKey: .sessionId)
        // Properties not encoded for simplicity
    }
}

/// Event categories
public enum EventCategory: String, Codable {
    case lifecycle
    case action
    case performance
    case error
    case feature
}

/// Metric value types
enum MetricValue {
    case count(Int64)
    case sum(Double)
    case average(sum: Double, count: Int64)
    case distribution([Double])
}

/// Analytics summary
public struct AnalyticsSummary {
    var totalEvents: Int = 0
    var eventsByCategory: [EventCategory: Int] = [:]
    var metrics: [String: String] = [:]
    var sessionDuration: Int = 0
}